wdl2cwl/main_oop.py

```diff
@@ -15,571 +14,331 @@
-    ram_min = f"$(inputs.{wdl_runtime_cpu})"
-    requirements.append(
-        cwl.ResourceRequirement(
-            ramMin=ram_min,
+            elif "sep" in options:
+                cwl_command_str = (
+                    f"$(inputs.{placeholder_name}.map("
+                    + 'function(el) {return el.path}).join("'
+                    + seperator
+                    + '"))'
+                )
+        elif isinstance(wdl_placeholder.expr, WDL.Expr.Apply):
+            function_name = wdl_placeholder.expr.function_name
+            expr_arguments = wdl_placeholder.expr.arguments
+
+            if function_name == "defined":
+                arg = expr_arguments[0]
+                if not isinstance(arg, WDL.Expr.Get):
+                    raise Exception(f"Unsupported type: {type(arg)}: {arg}")
+                arg_expr = arg.expr
+                if not isinstance(arg_expr, WDL.Expr.Ident):
+                    raise Exception(f"Unsupported type: {type(arg_expr)}: {arg_expr}")
+                arg_referee = arg_expr.referee
+                if not isinstance(arg_referee, WDL.Tree.Decl):
+                    raise Exception(
+                        f"Unsupported type: {type(arg_referee)}: {arg_referee}"
+                    )
+                arg_referee_name = arg_referee.name
+                cwl_command_str = (
+                    "$(inputs."
+                    + arg_referee_name
+                    + " ? "
+                    + f'"{false_value}" : "{true_value}")'
+                )
+            elif function_name == "_interpolation_add":
+                arg_name_raw, arg_value_raw = expr_arguments
+                arg_name_literal = arg_name_raw.literal
+                if arg_name_literal is None or not hasattr(arg_name_literal, "value"):
+                    raise Exception(f"Unsupported type: {type(arg_name_literal)}")
+                if not isinstance(arg_value_raw, WDL.Expr.Get):
+                    raise Exception(f"Unsupported type: {type(arg_value_raw)}")
+                arg_value_expr = arg_value_raw.expr
+                if not isinstance(arg_value_expr, WDL.Expr.Ident):
+                    raise Exception(f"Unsupported type: {type(arg_value_expr)}")
+                arg_name, arg_value = arg_name_literal.value, arg_value_expr.name
+                if wdl_placeholder.expr.type.optional:
+                    cwl_command_str = (
+                        "$(inputs."
+                        + arg_value
+                        + ' === null ? "" : '
+                        + f'"{arg_name}"'
+                        + " inputs."
+                        + arg_value
+                        + " )"
+                    )
+                else:
+                    cwl_command_str = f"{arg_name} $(inputs.{arg_value})"
+            elif function_name == "sub":
+                wdl_apply_object_arg, arg_string, arg_substitute = expr_arguments
+                if not hasattr(wdl_apply_object_arg, "arguments"):
```

Comment:
What type of `WDL` object from miniwdl has an attribute named `arguments`? Can we test for that type instead?

Reply:
That will be WDL.Expr.Apply